compiler-core/src/type_/pipe.rs

```diff
@@ -99,6 +103,25 @@ impl<'a, 'b, 'c> PipeTyper<'a, 'b, 'c> {
                     }
                 }
 
+                // If the left type has the correct number of arguments to the right function,
+                // we can infer the types of the arguments in the right function.
+                //
+                // left |> fn(arg) {} as fn(left)
+                UntypedExpr::Fn {
+                    arguments,
+                    body,
+                    return_annotation,
+                    is_capture,
+                    location,
+                } if arguments.len() == expected_args.len() => self.infer_apply_to_fn_pipe(
+                    arguments,
+                    &expected_args,
+                    body,
+                    is_capture,
+                    &return_annotation,
+                    location,
+                )?,
```

Comment:
I think this would be best done in the context of function application, not in the sugar rewriting code here. It would keep this code clearer, and it would mean that this works outside of pipes too.

```gleam
  let z = #(1,2)
  fn(x) { x.0 }(z)
```

Reply:
Thanks, looking into this!